### PR TITLE
support for openflow 1.3

### PR DESCRIPTION
**Pull request by nikhil1290**
Adding support for ```OpenFlow 1.3``` and support for merging of prefixes

### CODE REVIEW

File: python/Ryu.py [hunk 1]
```diff
@@ -199,12 +200,38 @@ def changeSwitchForwardingState(self, dpid=None, header=None, actions=None, comm
...
 
+    def OF10_flow(self, dp=None, domain=None, header=None, actions=None, command=None, idle_timeout=None, hard_timeout=None, priority=1):
+        ofp      = dp.ofproto
+        parser   = dp.ofproto_parser
         obj = {} 
         
         if(header.has_key('dl_type')):
             if(header['dl_type'] == None):
                 obj['dl_type'] = None
+                del header['dl_type']
```
Comment (ajragusa) [inline python/Ryu.py]: Why do we need to delete the header dl_type?

File: python/Ryu.py [hunk 2]
```diff
@@ -216,23 +243,21 @@ def changeSwitchForwardingState(self, dpid=None, header=None, actions=None, comm
             obj['in_port'] = None
             
         if(header.has_key('nw_src')):
+            if header['nw_src'].version == 6:
+                obj['dl_type'] = 34525
```
Comment (ajragusa) [inline python/Ryu.py]: Seems like there needs to be an else here???  OF 1.0 can't include IPv6 maybe Ryu is doing something special for us?  In any case I think it is safer to have an else so that we don't try and set the nw_src/dst for IPv6 in OF1.0

File: python/Ryu.py [hunk 3]
```diff
@@ -216,23 +243,21 @@ def changeSwitchForwardingState(self, dpid=None, header=None, actions=None, comm
...
             obj['nw_src'] = None
-             
-        if(header.has_key('nw_src_mask')):
-            obj['nw_src_mask'] = int(header['nw_src_mask'])
-        else:
             obj['nw_src_mask'] = None
          
         if(header.has_key('nw_dst')):
+            if header['nw_dst'].version == 6:
+                obj['dl_type'] = 34525
```
Comment (ajragusa) [inline python/Ryu.py]: Same as for lines 246 and 247

File: python/Ryu.py [hunk 4]
```diff
@@ -273,31 +282,247 @@ def changeSwitchForwardingState(self, dpid=None, header=None, actions=None, comm
...
+        
+        if(header.has_key('phys_port')):
+            obj['in_port'] = int(header['phys_port'])
+            header['in_port'] = int(header['phys_port'])
+            del header['phys_port']
+
+        if(header.has_key('nw_src')):
+            if(header['nw_src'].version == 4):
+                obj['ipv4_src'] = (str(header['nw_src'].ip), str(header['nw_src'].netmask))
+                obj['eth_type'] = 0x800
```
Comment (ajragusa) [inline python/Ryu.py]: can we use an existing ENUM or create a new one so that we don't have to know IPv4 ethernet type is 0x800 and IPv6 is 0x86dd?

File: python/Ryu.py [hunk 5]
```diff
@@ -273,31 +282,247 @@ def changeSwitchForwardingState(self, dpid=None, header=None, actions=None, comm
...
+            else:
+                del obj['ipv4_src']
+            if(header['nw_src'].version == 6):
+                if(obj.has_key('ipv4_src')):
+                    self.logger.error("IPv4 and IPv6 in the same message")
+                    return
+                obj['ipv6_src'] = (str(header['nw_src'].ip), str(header['nw_src'].netmask))
+                obj['eth_type'] = 0x86dd
+            else:
+                del obj['ipv6_src']
```
Comment (ajragusa) [inline python/Ryu.py]: Why is this here?

File: python/Ryu.py [hunk 6]
```diff
@@ -273,31 +282,247 @@ def changeSwitchForwardingState(self, dpid=None, header=None, actions=None, comm
...
         else:
-            self.logger.error("Device is not connected")
+            del obj['ipv4_src']
+            del obj['ipv6_src']
+
```
Comment (ajragusa) [inline python/Ryu.py]: same for these 2

File: python/Ryu.py [hunk 7]
```diff
@@ -273,31 +282,247 @@ def changeSwitchForwardingState(self, dpid=None, header=None, actions=None, comm
...
+        if(command == "ADD"):
+            command = ofp.OFPFC_ADD
+            flags = ofp.OFPFF_SEND_FLOW_REM
+            self.api.pushTimeouts(idle, hard)
+            self.api.pushFlows(dpid, domain, header, actions, priority)
+        elif(command == "DELETE_STRICT"):
+            command = ofp.OFPFC_DELETE_STRICT
+            self.api.remove_flow(dpid, domain, header, priority)
+        else:
+            command = ofp.OFPFC_DELETE
```
Comment (ajragusa) [inline python/Ryu.py]: Why would we only want to delete instead of delete_strict?  Seems like we could accidentally delete a lot of flows?

File: python/Ryu.py [hunk 8]
```diff
@@ -482,33 +760,285 @@ def _flow_stats_reply_handler(self, ev):
...
+                
+            return count
+
+        def ipv6Prefix(addr):
+            """returns IPv6 object for a given address and netmask"""
+            prefix = None
+            if type(addr) is tuple:
+                mask = getCIDR(str(addr[1]))
+                if mask == 0:
+                    prefix = ipaddr.IPv4Network(str(addr[0]))
```
Comment (ajragusa) [inline python/Ryu.py]: I presume this is suppose to be IPv6Network instead of IPv4Network?

File: python/Ryu.py [hunk 9]
```diff
@@ -564,8 +1094,29 @@ def process_flow_stats(self, stats, dp):
             wildcards = stat.match.wildcards
             del match['dl_dst']
             del match['dl_src']
-            del match['dl_type']
-            del match['wildcards']
+            del match['wildcards'] 
+            if match['dl_type'] == 34525:
+                if src_mask > 0:
```
Comment (ajragusa) [inline python/Ryu.py]: we can't match on IPv6 in 1.0 so I guess I don't fully understand what this is attempting to do?  Can you explain?

File: python/SciPass.py [hunk 10]
```diff
@@ -922,14 +947,13 @@ def addPrefix(self, dpid=None, domain_name=None, group_id=None, prefix=None, pri
                                               idle_timeout = 0,
                                               hard_timeout = 0,
                                               priority     = priority)
-
-
+      
       if(prefix._version != 4):
-        header = {"dl_type": 34525}
+        header = {"dl_type": 34525,
```
Comment (ajragusa) [inline python/SciPass.py]: can we remove this if? seems like this is taken care of based on the prefix type later on?

File: python/Ryu.py (hunk 8 above)
Comment (nikhil1290) [inline python/Ryu.py]: Agree!

File: python/Ryu.py (hunk 9 above)
Comment (nikhil1290) [inline python/Ryu.py]: Agree.I think, I jumbled that with the 1.3 stuff.I'll fix this.

File: python/t/etc/Simple_SciPass_balancer_only.xml [hunk 11]
```diff
@@ -3,6 +3,7 @@
    <switch dpid="0000000000000001" name="IUPUI-SW">
     <domain name="IUPUI" mode="SimpleBalancer" admin_status="active" least_specific_prefix_len="16" most_specific_prefix_len="24" 
 	    ignore_sensor_load="true" blacklist_priority="1000" whitelist_priority="900" sensor_min_load_threshold="0.1" 
+	    ipv6least_specific_prefix_len="64" ipv6most_specific_prefix_len="128" sensor_configurable_threshold="100"
```
Comment (ajragusa) [inline python/t/etc/Simple_SciPass_balancer_only.xml]: these values for IPv6 seem wrong... I would think the most specific would be /64 (there are only a few quadrillion of them... I guess I'm not sure how this actually could work seems like if you included anything you would have essentially infinite addresses?

File: python/SimpleBalancer.py [hunk 12]
```diff
@@ -714,7 +744,78 @@ def getEstLoad(self,group,targetPrefix):
...
+                  aggBW = aggBW*1000*1000
+                  self.addGroupPrefix(minSensor, candidatePrefix, aggBW)
+              except DuplicatePrefixError:
+                  self.logger.debug("Already have prefix: " + str(prefix))
+                  return
+              except MaxFlowCountError:
+                  self.logger.debug("Max Flow Count Error")
+                  return
+              return
+
```
Comment (ajragusa) [inline python/SimpleBalancer.py]: seems like if we hit an exception we need to add the group prefixes back, otherwise we might loose prefixes we were balancing on

**Comment by nikhil1290** [comment]
Fixed the above specified bugs/issues.

File: python/SimpleBalancer.py [hunk 13]
```diff
@@ -714,7 +745,83 @@ def getEstLoad(self,group,targetPrefix):
...
+                  self.addGroupPrefix(group_a, prefix_a, bw1)
+                  self.addGroupPrefix(group_b, prefix_b, bw2)
+                  return
+              except MaxFlowCountError:
+                  self.logger.debug("Max Flow Count Error")
+                  self.addGroupPrefix(group_a, prefix_a, bw1)
+                  self.addGroupPrefix(group_b, prefix_b, bw2)
+                  return
+              return
+
```
Comment (ajragusa) [inline python/SimpleBalancer.py]: So I think a better solution here is to delete the individual prefixes after we have successfully added the merged prefix.  That should simplify this code a little bit and prevent packet loss!

**Comment by ajragusa** [comment]
Hey Nikhil, Great job!
